### DISCUSSION TITLE
Generating Tables with Julia

### DISCUSSION THREAD

**Discussion by gdietl** [quarto-dev/quarto-cli#4522] [Q&A] [answered]
With quarto, it is possible to  generate tables with R or Python. Is there also a way to generate tables with Julia? I've tried using PrettyTables.jl and producing an output in markdown format. However, the output is not correctly interpreted by Quarto, even note wenn setting "output: asis".

**Comment by gdietl** [comment]
I found a solution. It does not work with PrettyTables.jl but with Tables.jl and MarkdownTables.jl. It follows a code example:
```
```{julia}
#| label: tbl-example
#| tbl-cap: Example
#| echo: false
using Markdown,Tables,MarkdownTables
data = ["A" 60; "B" 40; "C" 50];
# Some calculations (not important for this example)
result = [data; "Total" sum(data[1:end,2])];
markdown_table(Tables.table(result))
```
```

**Reply by gdietl** [reply]
Note that this solution only works when quarto is using Jupyter. If there are R-chunks in the document and quarto executes Julia via R, the table is not generated.

**Reply by cderv** [reply, marked as answer]
I think this is because `JuliaCall` R package does not handle the default output of `markdown_table` which is a [Mime type](https://github.com/tpapp/MarkdownTables.jl/blob/master/src/MarkdownTables.jl#L70)  outputing [`DisplayAs.Raw.Showable`](https://tkf.github.io/DisplayAs.jl/dev/#DisplayAs.Raw.Showable) object. 

[JuliaCall](https://github.com/Non-Contradiction/JuliaCall) is the package called by **knitr** when Julia chunk are used, and knitr handles the output based on the method defined by this tool. 

You can make the above works with **knitr** by outputing a Markdown table as string and using `output: asis`
````markdown
---
title: "Tables"
format: html
keep-md: true
engine: knitr
---

```{julia}
#| label: tbl-example
#| tbl-cap: Example
#| output: asis
using Markdown,Tables,MarkdownTables
data = ["A" 60; "B" 40; "C" 50];
# Some calculations (not important for this example)
result = [data; "Total" sum(data[1:end,2])];
print(markdown_table(Tables.table(result), String))
```
````

Jupyter or IJulia probably handles the `DisplayAs.Raw.Showable` type directly and we get a Jupyter output that can be handled by Quarto 

Hope it helps understand.

**Reply by gdietl** [reply]
Unfortunately, your code is not giving me the right output. As is, it still uses Jupyter and procudes the following output:
![Screenshot_20230412_162634](https://user-images.githubusercontent.com/112486210/231488992-18460198-36d2-40df-921f-063f420ebe9d.png)

If I define knitr as the engine by using the following code
````
---
title: "Tables"
format: html
keep-md: true
engine: knitr
---

```{julia}
#| label: tbl-example
#| tbl-cap: Example
#| output: asis
using Markdown,Tables,MarkdownTables
data = ["A" 60; "B" 40; "C" 50];
# Some calculations (not important for this example)
result = [data; "Total" sum(data[1:end,2])];
markdown_table(Tables.table(result), String)
```
````
it is using JuliaCall vom knitr and I'll get the output:
![Screenshot_20230412_162843](https://user-images.githubusercontent.com/112486210/231489610-7ddf4231-e377-491f-8a0a-20922356e754.png)

The markdown string is correct in this case but it is still not interpreted correctly. Do I miss something?

Besides, when producing a pdf document via xelatex, I'll get the error
```
compilation failed- missing packages (automatic installed disabled)
Undefined control sequence.
<argument> ...1 \textbar {} Column2 \textbar {}\n 
                                                  \textbar ---------\textbar...
l.214 \textbar{} 150 \textbar{}\n''}
```

**Reply by cderv** [reply]
Sorry I missed a line in the YAML header while copy pasting and cleaning my example. You need to add `engine: knitr` as you have tried in YAML if the `julia` chunk is the only one and no other R chunk is inside the qmd document. Otherwise Jupyter will be used. 

And I mixed up my copy paste - with the String format, an explicit `print()` will be needed for the `\n` to be printed correctly. 

I edited the example above. Sorry for that.

**Reply by gdietl** [reply]
Now, it is working. Thanks for your helpful solution! I've marked the above comment from you as the answer.

**Comment by DelmerBratzOd2** [comment]
Yes, it is definitely possible to generate tables with Julia in Quarto. While PrettyTables.jl is a great package for creating tables in Julia, it might not work directly with Quarto's Markdown output.

Instead, I would recommend using the DataFrames.jl package to create tables in Julia, and then converting them to a format that can be easily imported into Quarto. One way to do this is to use the CSV.jl package to write the table to a CSV file, and then import the CSV file into Quarto.

Here's an example of how you can create a table using DataFrames.jl, write it to a CSV file, and then import it into Quarto:D
![image](https://user-images.githubusercontent.com/127719006/224621321-15511163-8113-4db0-8e2f-17c451331981.png)
Then, in your Quarto document, you can import the table using the following Markdown syntax:
![image](https://user-images.githubusercontent.com/127719006/224621379-97e9df5e-13a8-4d8b-8e2d-c551bb779897.png)

**Reply by gdietl** [reply]
Thanks for your solution. I did not know that you can directly import CSV files into quarto. This might be the perfect solution in cases where Julia is executed via R.

**Reply by gdietl** [reply]
I've tried using your code but, unfortunately, it is not working. The table-chunk is not recognized. I'm running quarto version 1.2.280. Do I have to install a new version?

**Reply by cderv** [reply]
@DelmerBratzOd2 curious how this `table` chunk is working ? Is this from a specific extension ? 

Thanks

**Comment by WarnerMarkakisA** [comment]
One option is to use the DataFrames.jl package to create a data frame in Julia, and then use the package's showtable() function to display the data frame as a table. You can then include this code chunk in your Quarto document and set the output format to "asis" to ensure that the table is correctly rendered.

Another option is to use the CSV.jl package to read in a CSV file with your data, and then use the DataFrames.jl package to manipulate and display the data as a table. Again, you can include this code chunk in your Quarto document and set the output format to "asis" to ensure that the table is correctly rendered.

Here's an example code chunk that demonstrates how to use DataFrames.jl to create a table in Julia:
using DataFrames

# create a data frame
df = DataFrame(A = [1, 2, 3], B = [4, 5, 6], C = [7, 8, 9])

# display the data frame as a table
showtable(df)

**Reply by agucova** [reply]
Is this a GPT response? `showtable` doesn't exist.

**Reply by cderv** [reply]
Thanks @agucova - This is highly possible considering the profile of the user. I'll hide this answer
